### ISSUE TITLE
Self-shooting still ocurrs very often with latest SVN

### ISSUE THREAD

**Issue by albertz** [openlierox/openlierox#326]
Converted from [SourceForge issue 2811445](http://sourceforge.net/support/tracker.php?aid=2811445), submitted by pelya

I'm using latest OLX (r4650 currently), and regularly shoot myself with weapons like Doomsday or Cannon in Classic mod, and many others. This gives big advantage to LX56 users over me, because self-shooting doesn't ever occurs for LX56 (as reported by Cizin).
My ping is 90-100 and pretty stable, so I guess our previous attempts to fix self-shooting failed (have we ever tested them in real network conditions?).
I'm proposing following solution - when server gives us projectiles list of our own worm, spawn them not where server said, but in front of our worm. Well of course there will be sanity check that it won't affect projectiles spawned away from worm, but only those whose coordinates match our worm coordinates halfsecond ago.
And we can remove that hack that your own projectile won't damage you for 100 milliseconds or so - it doesn't work anyway.
That will desync physics slightly, but self-shooting is way worse.

**Comment by albertz** [comment]
Submitted by albertzeyer

Please, before just trying something, check why it is not working right now. I think, the current solution is still the best. And if it is not working, there must be a bug which hopefully can be fixed.

Perhaps it's the time synchronisation, which is the base for the current system. When the client got the shoot info, it contains the info when in the past the shoot was done. The client then does the whole simulation of the projectile from that time to what it thinks what is "now", without doing worm collision check.

Or does it still do worm collision check but just doesn't make damage? That would also be a bug and also would explain the self-shooting.

**Comment by albertz** [comment]
Submitted by albertzeyer

Btw., I think your suggested solution is an even bigger hack. We really just should fix the problem itself, and not add hacky workarounds.

Your suggested solution also has a lot of other drawbacks. For example the projectile speed will be different than it would be (in some cases, it could even be a lot different).

**Comment by albertz** [comment]
Submitted by pelya

I've played on 0.56 server when last checked that, so no time synchronization was done.
And all my shots did injure my worm pretty bad.

No, the speed should stay the same here, if the offset of spawned projectile will be just 10-20 pixels it won't be that much noticeable as different speed.
Also I think offset shall be disabled if the projectile does not have worm-collide action (which should mean it is a grenade).

**Comment by albertz** [comment]
Submitted by albertzeyer

There are cases where the speed could have changed a lot, even with very low pings (it depends on the projectile; you easilly could make a projectile which increases/decreases the speed very fast). And with high pings, it could be very much different.

I think the time estimation for old clients is not that bad, but I am not sure. Is it big different if you play on LX56 or on Beta9 server?

However, have you checked the possible bug I noted?

I still think, if it would be correct (the current way), it should work as perfect as possible without any drawbacks. Or in a different way: Any other possible way would be worse (despite when we do just everything server-side).

**Comment by albertz** [comment]
Submitted by pelya

When I've increased the anti-self-shooting time coefficient, which is calculated from ping (was like ping / 2, made it ping \* 2), the self-shooting occurred much less, yet still occurred, the same way on LX56 and Beta8 servers - basically when you're shooting and there's small lagspike you shoot yourself, and only after that your ping updated to higher value.
Also that algorithm applied to all worms in game, so when someone shoots at you his bullets appear just before your worm if he aims correctly, harder to dodge - I think that thing should be used only for local worms, because in LX56 it was like that, and I don't want to give LX56 users any advantage :) .

I thought of another way - record your worm coordinates for half-second back, then compare them to the shootlist coordinates from server, and calculate the more precise time-offset, robust to lagspikes.

**Comment by albertz** [comment]
Submitted by albertzeyer

Please don't just change the coefficient to make self-shooting less. The coeffizient should be most exact as possible. The coeffizient should not work around another bug which is somewhere there in physics.

Also, why do you thinking so much about other complicated and ugly workarounds? Why aren't you trying to figure out what the real problem is? Self-shooting should already be impossible with current code.

**Comment by albertz** [comment]
Submitted by pelya

As far as I understand for pre-Beta8 server the client uses ping value, which may not be updated immediately as every new packet from server arrived, but for each second or third packet. As for Beta8+ it sends server time packet, and time inside each shot info packet, yet it's the same as ping algorithm, because between that two packets a lag may occur (and it does as far as I've seen).

**Comment by albertz** [comment]
Submitted by albertzeyer

The time it is sending is the TimeDiff to game start, and all the time try try to have the game time almost the same everywhere.

Anywhere, have you checked the possible bug I mentioned?

And before you are adding any workarounds here, please fix first the real problem and look, if it perhaps is working perfectly then.

**Comment by albertz** [comment]
Submitted by albertzeyer

Check r4693.

**Comment by albertz** [comment]
Submitted by pelya

I've tested it on Cizin's server, he turned on torrents and I've got unstable 100-150 ping.
I've managed to shoot myself few times, though it was really hard.
Cizin also said that he shoot himself earlier when he was hosting (eek how's that even possible) but now it's gone.
So I guess there's no need for writing advanced anti-self-shooting system which calculates server time diff by distance between your worm and spawned projectile.
